File: app/src/main/java/com/czbix/v2ex/util/TrackerUtils.java
package com.czbix.v2ex.util;

import android.content.Context;

import com.czbix.v2ex.BuildConfig;
import com.czbix.v2ex.R;
import com.google.android.gms.analytics.GoogleAnalytics;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Logger;
import com.google.android.gms.analytics.Tracker;
import com.google.common.base.Preconditions;

import java.util.Map;

public class TrackerUtils {
    private static GoogleAnalytics mAnalytics;
    private static Tracker mTracker;

    public static void init(Context context) {
        Preconditions.checkState(mAnalytics == null, "can't init twice");

        mAnalytics = GoogleAnalytics.getInstance(context);
        if (BuildConfig.DEBUG) {
            mAnalytics.getLogger().setLogLevel(Logger.LogLevel.VERBOSE);
            mAnalytics.setDryRun(true);
        }

        mTracker = mAnalytics.newTracker(R.xml.analytics_config);

        onAppCreate();
    }

    public static void onAppCreate() {
        final Map<String, String> hit = new HitBuilders.EventBuilder(Category.APP, Action.CREATE).build();
        mTracker.send(hit);
    }

    public static void onTopicSwitchReply(boolean isShow) {
        final Map<String, String> hit = new HitBuilders.EventBuilder(Category.TOPIC, Action.SWITCH_REPLY)
                .setLabel(isShow ? Label.SHOW : Label.HIDE).build();
        mTracker.send(hit);
    }

    public static void onTopicReply() {
        final Map<String, String> hit = new HitBuilders.EventBuilder(Category.TOPIC, Action.REPLY).build();
        mTracker.send(hit);
    }

    public static void onSearch() {
        final Map<String, String> hit = new HitBuilders.EventBuilder(Category.APP, Action.SEARCH).build();
        mTracker.send(hit);
    }

    private static class Category {
        public static final String APP = "App";
        public static final String TOPIC = "Topic";
    }

    private static class Action {
        public static final String CREATE = "Create";
        public static final String SWITCH_REPLY = "Switch Reply";
        public static final String REPLY = "Reply";
        public static final String SEARCH = "Search";
    }

    private static class Label {
        public static final String SHOW = "Show";
        public static final String HIDE = "Hide";
    }
}

File: app/src/main/java/com/czbix/v2ex/util/TrackerUtils.kt
package com.czbix.v2ex.util

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import java.security.MessageDigest

object TrackerUtils {
    private lateinit var analytics: FirebaseAnalytics

    private fun hashString(data: String): String {
        val salted = "cz-%s-bix".format(data)
        return MessageDigest.getInstance("MD5")
                .digest(salted.toByteArray())
                .fold("") { str, it ->
                        str + "%02x".format(it)
                }
    }

    fun init(context: Context) {
        analytics = FirebaseAnalytics.getInstance(context)
    }

    fun setUserId(id: String?) {
        val hashedId = if (id == null) null else hashString(id)
        analytics.setUserId(hashedId)
    }

    fun onTopicSwitchReply(isShow: Boolean) {
        val params = Bundle().apply {
            putBoolean("is_show", isShow)
        }

        analytics.logEvent(Event.SWITCH_REPLY, params)
    }

    fun onTopicReply() {
        analytics.logEvent(Event.REPLY_TOPIC, null)
    }

    fun onSearch() {
        analytics.logEvent(FirebaseAnalytics.Event.SEARCH, null)
    }

    fun onParseTopic(time: Long, commentCount: Int) {
        val params = Bundle().apply {
            putInt("cost_time", time.toInt())
            putInt("topic_count", commentCount)
        }
        analytics.logEvent(Event.PARSE_TOPIC, params)
    }

    private object Event {
        const val SWITCH_REPLY = "switch_reply"
        const val REPLY_TOPIC = "reply_topic"
        const val PARSE_TOPIC = "parse_topic"
    }
}